packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1417,6 +1417,90 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
+  /// the `delegatedTransition` from the incoming transition to animate off the
+  /// screen.
+  /// {@endtemplate}
+  ///
+  /// The [ModalRoute] receiving this transition will set it to their
+  /// [receivedTransition] property.
+  ///
+  /// {@tool dartpad}
+  /// This sample shows an app that uses three different page transitions, a
+  /// Material Zoom transition, the standard Cupertino sliding transition, and a
+  /// custom vertical transition. All of the page routes are able to inform the
+  /// previous page how to transition off the screen to sync with the new page.
+  ///
+  /// ** See code in examples/api/lib/widgets/routes/flexible_route_transitions.0.dart **
+  /// {@end-tool}
+  ///
+  /// {@tool dartpad}
+  /// This sample shows an app that uses the same transitions as the previous
+  /// sample, this time in a [MaterialApp.router].
+  ///
+  /// ** See code in examples/api/lib/widgets/routes/flexible_route_transitions.1.dart **
+  /// {@end-tool}
+  DelegatedTransitionBuilder? get delegatedTransition => null;
+
+  /// The [DelegatedTransitionBuilder] received from the route above this one in
+  /// the navigation stack.
+  ///
+  /// {@macro flutter.widgets.delegatedTransition}
+  ///
+  /// `receivedTransition` will use the above route's [delegatedTransition] in
+  /// order to show the right route transition when the above route either enters
+  /// or leaves the navigation stack. If not null, the `receivedTransition` will
+  /// wrap the route content through [buildFlexTransitions].
+  DelegatedTransitionBuilder? receivedTransition;
+
+  /// Wraps the transitions of this route with a [DelegatedTransitionBuilder], when
+  /// [receivedTransition] is not null.
+  ///
+  /// This allows the current route to sync its exit transition with the
+  /// entrance transition of the incoming route, as well as the exit animation
+  /// when the above route is popped.
+  ///
+  /// {@tool dartpad}
+  /// This sample shows an app that uses three different page transitions, a
+  /// Material Zoom transition, the standard Cupertino sliding transition, and a
+  /// custom vertical transition. All of the page routes are able to inform the
+  /// previous page how to transition off the screen to sync with the new page.
+  ///
+  /// ** See code in examples/api/lib/widgets/routes/flexible_route_transitions.0.dart **
+  /// {@end-tool}
+  ///
+  /// {@tool dartpad}
+  /// This sample shows an app that uses the same transitions as the previous
+  /// sample, this time in a [MaterialApp.router].
+  ///
+  /// ** See code in examples/api/lib/widgets/routes/flexible_route_transitions.1.dart **
+  /// {@end-tool}
+  Widget buildFlexTransitions(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    final Widget originalTransitions = buildTransitions(context, animation, secondaryAnimation, child);
+
+    if (receivedTransition == null) {
+      return originalTransitions;
```

Comment:
Is there any benefit to avoiding building originalTransitions in the cases where they are not needed? I guess building a widget tree is quick so it shouldn't be a big deal...